### PR TITLE
Supply mule should now take the actual amount that could be bought by…

### LINKED ISSUE #46: supply mule performance issue when source station runs out of money
When a station runs out of money and subordinate supply mule wants to make a supply run at a NPC station the script runs into and endless loop and basically crashes performance.
Steps to reproduce:
1) Setup a new production station without any budget (or let it run out)
2) Make sure you dont have any player station producing the required goods
3) Assign a supply mule to the station using the subordinate feature (checkbox: Assign Ship To Station)
4) Watch your fps drop.

In the error log you can see the failsave from egosoft kicking in:
```[Scripts] 175006.05 *** aicontext<order.trade.perform,0x3aa02>: 175006.054 'Supply Mule (Claytronics Factory Alpha)'(job: '') TRADE RUN FAILED! Error: ERR_NO_CARGO_SPACE
[General] 175006.05 ======================================
[=ERROR=] 175006.05 Prevented account 0x110 from going negative, see callstack [ Michael ]
[General] 175006.05 ======================================
```

The debug log from the mule:

```
ScriptStart
found subordinates: 0
 Setting subordinates used up: 0 Seconds
 Selling leftover cargo used up: 0 Seconds
Define Wares
dedicated serve: 1
sourceStation: 0x3aa2c
----- Source Station set to Claytronics Factory Alpha. Entering Source Station owned by Player block 
----- player supply to station build storage
----- station build needs
stationBuildNeeds count: 0
station build storage needs:
----- player owned supply offers
playerSupplyOffers count: 0
player Supply Offers:
----- ai to station build storage
----- station build needs
stationBuildNeeds count: 0
station build storage needs:
----- ai supply offers
aiSupplyOffers count: 7
ai Supply Offers:
    FRF Quantum Tube Factory I Quantum Tubes 2272 -0.487333 0 0.6497
    FRF Microchip Factory II Microchips 2425 -0.884545 0 0.846422
    FRF Quantum Tube Factory I Quantum Tubes 5762 -0.833733 0 0.821149
    TEL Microchip Factory I Microchips 4874 -0.95028 0 0.841506
    FRF Quantum Tube Factory I Quantum Tubes 2374 -0.557467 0 0.676739
    FRF Microchip Factory I Microchips 7847 -1 0 0.482981
    FRF Antimatter Cell Factory I Antimatter Cells 52492 -1 0 0.989276
----- player to station production goods
----- station product needs
stationProductNeeds count: 3
station product storage needs:
    Claytronics Factory Alpha Antimatter Cells 0 0.99925 7046 0
    Claytronics Factory Alpha Microchips 0 0.505775 7534 0
    Claytronics Factory Alpha Quantum Tubes 0 1 7462 0
----- player supply offers
playerSupplyOffers count: 0
player Supply Offers:
----- anyone to station production goods
----- station product needs
stationProductNeeds count: 3
station product storage needs:
    Claytronics Factory Alpha Antimatter Cells 0 0.99925 7046 0
    Claytronics Factory Alpha Microchips 0 0.505775 7534 0
    Claytronics Factory Alpha Quantum Tubes 0 1 7462 0
----- ai supply offers
aiSupplyOffers count: 7
ai Supply Offers:
    FRF Quantum Tube Factory I Quantum Tubes 2272 -0.487333 0 0.6497
    FRF Microchip Factory II Microchips 2425 -0.884545 0 0.846422
    FRF Quantum Tube Factory I Quantum Tubes 5762 -0.833733 0 0.821149
    TEL Microchip Factory I Microchips 4874 -0.95028 0 0.841506
    FRF Quantum Tube Factory I Quantum Tubes 2374 -0.557467 0 0.676739
    FRF Microchip Factory I Microchips 7847 -1 0 0.482981
    FRF Antimatter Cell Factory I Antimatter Cells 52492 -1 0 0.989276
ship capacity: 7500
Occupied Cargo: 0
----- trade offer comparisons
Antimatter Cells target desired amount 7046 supply amount: 52492 ship can hold: 416 amount will be: 416 need unitprice: 28194 supp unitprice: 12100 player buy modifier: 0 profit: 6695104
Microchips target desired amount 7534 supply amount: 2425 ship can hold: 340 amount will be: 340 need unitprice: 101982 supp unitprice: 82151 player buy modifier: 0 profit: 6742540
Microchips target desired amount 7534 supply amount: 4874 ship can hold: 340 amount will be: 340 need unitprice: 101982 supp unitprice: 81211 player buy modifier: 0 profit: 7062140
Microchips target desired amount 7534 supply amount: 7847 ship can hold: 340 amount will be: 340 need unitprice: 101982 supp unitprice: 80500 player buy modifier: 0 profit: 7303880
Quantum Tubes target desired amount 7462 supply amount: 2272 ship can hold: 340 amount will be: 340 need unitprice: 37500 supp unitprice: 26345 player buy modifier: 0 profit: 3792700
Quantum Tubes target desired amount 7462 supply amount: 5762 ship can hold: 340 amount will be: 340 need unitprice: 37500 supp unitprice: 23747 player buy modifier: 0 profit: 4676020
Quantum Tubes target desired amount 7462 supply amount: 2374 ship can hold: 340 amount will be: 340 need unitprice: 37500 supp unitprice: 25819 player buy modifier: 0 profit: 3971540
buying 340 Microchips from FRF Microchip Factory I at 805 Cr to sell to Claytronics Factory Alpha at 1,019 Cr for a profit of 73,038 Cr
Occupied Cargo: 7480
Cargo Full, No More Trades
******* went to end
```

The fix would require the mule to take the stations balance into account if it is a subordinate and trading with NPC stations.
I will try to provide a patch later today.
For now the workaround would be to make sure the station has enough funds.

### PR DESCRIPTION
**Pull request by Galaxy-Spam**
… the station into account when buying from NPCs

partically fixes #46

### CODE REVIEW

**Comment by bhayden53** [comment]
Were you able to test this in your save?

**Comment by Galaxy-Spam** [comment]
yes, did test it before, it worked out (git repo is in extensions folder).
However it only fixes a single case!

**Comment by bhayden53** [comment]
Yep, I know, I will go through and edit all trades with the AI to also check the affordable amount. It will be pushed to this PR and then merged. Thanks for all the help.

**Comment by Galaxy-Spam** [comment]
Thank you for the mod ;)

**Comment by Misunderstood-Wookiee** [comment]
Suppose I'd put out a new release later today then.

**Comment by bhayden53** [comment]
I don't think there's going to be a release ready today, I have a lot of work I need to do to get a few things ready. Nothing game breaking has shown up IMO so I don't see a need to go through the annoyance of putting out a release. 

There's no specific issue that is yet fully fixed anyway. But I am working on the issues that have piled up this weekend, so we'll see how far I get.

File: aiscripts/supply_mule.xml [hunk 1]
```diff
@@ -5113,11 +5116,11 @@
...
-									<debug_to_file chance="$debugchance" name="'SupplyMule- ' + this.ship.knownname + ' ' + this.ship.idcode" directory="'MulesExtended'" text="'buying ' +$bestAmount + ' ' +$bestNeedTrade.ware +' from ' +$bestSupplyTrade.owner.knownname
+									<debug_to_file chance="$debugchance" name="$debugFileName" directory="'MulesExtended'" text="'buying ' +$bestAmount + ' ' +$bestNeedTrade.ware +' from ' +$bestSupplyTrade.owner.knownname
 									+' at ' +$bestSupplyTrade.unitprice.formatted.{'%s %Cr'} + ' to sell to ' +$bestNeedTrade.owner.knownname
 									+' at ' +$bestNeedTrade.unitprice.formatted.{'%s %Cr'} + ' for a profit of ' +$bestProfit.formatted.{'%s %Cr'} " />
 	
-									<write_to_logbook category="upkeep" title="'SupplyMule- ' + this.ship.knownname + ' ' + this.ship.idcode" interaction="showonmap" money="$bestProfit" text="'buying ' +$bestAmount + ' ' +$bestNeedTrade.ware +' from ' +$bestSupplyTrade.owner.knownname
+									<write_to_logbook category="upkeep" title="$debugFileName" interaction="showonmap" money="$bestProfit" text="'buying ' +$bestAmount + ' ' +$bestNeedTrade.ware +' from ' +$bestSupplyTrade.owner.knownname
```
Comment (bhayden53) [inline aiscripts/supply_mule.xml]: We may actually want the ship name in the logbook entries.

Comment (Galaxy-Spam) [reply aiscripts/supply_mule.xml]: Ah, mass replace... Let me check that

Comment (Galaxy-Spam) [reply aiscripts/supply_mule.xml]: should be seperated now, see commit https://github.com/Misunderstood-Wookiee/Mules-and-Warehouses-Extended/pull/47/commits/d30d289c8f7b4d402fc69a274b3163de21e18b5e

**Comment by bhayden53** [comment]
I found the mod that was breaking my dev environment so I've got this running in my save now. Want to keep an eye on it for a little while before merging. Just want to see it run for a couple of hours and check the logs. Thanks for showing me their string formatting and also for clamp_trade_amount. There's probably a lot of convenience functions that could replace alot of code. One thing at a time though...

**Comment by Galaxy-Spam** [comment]
yeah, no problem. I just started modding X4 when this problem showed up for me (3 days ago), so i had to learn all that stuff xD.

**Comment by bhayden53** [comment]
I just started a couple of weeks ago myself, I apparently didn't take enough time to explore the functions :)

**Comment by bhayden53** [comment]
This looks fine to me, only issues in my log from running overnight are other things that I know about.

**Pull request merged**